ring link changes

TheBurgerTV2 opened the pull request:
you can ignore the changes i made to the readme i wasnt even sure if i was gonna pr this lol

i will say first hand that most of this was written by studkid, the dev of ror1 apworld. i did most of the testing and the actual implementing the code

the only issue i have had is that i cannot patch properly on 0.6.0rc1, though a friend was able to with no issues. so i think my install is just bugged

TheBurgerTV2 committed:
Update README.md

TheBurgerTV2 committed:
Update README.md

TheBurgerTV2 committed:
ring link code

TheBurgerTV2 committed:
ring link code

TheBurgerTV2 committed:
more ring link code

TheBurgerTV2 committed:
ring link option

TheBurgerTV2 committed:
ring link

TheBurgerTV2 committed:
ring link code

Cyb3RGER commented on BfBBClient.py:
```diff
@@ -596,6 +601,8 @@ def on_package(self, cmd: str, args: dict):
             if 'death_link' in args['slot_data']:
                 Utils.async_start(self.update_death_link(bool(args['slot_data']['death_link'])))
+            if 'ring_link' in args['slot_data']:
+                self.ring_link = args['slot_data']["ring_link"]
```
I would prefer to have a function ``update_ring_link`` here, like the ``CommonClient.update_death_link``, instead of changing/checking tags in ``handle_ring_link``.

TheBurgerTV2 replied:
thats fine. i'm not sure how to do that though, if you can send me the rewrite i can implement it

Cyb3RGER commented on BfBBClient.py:
```diff
@@ -582,6 +582,11 @@ def __init__(self, server_address, password):
         self.current_scene_key = None
+        self.ring_link = False
+        self.previous_rings = None
+        self.ring_link_rings = 0
+        self.game_tags = []
```
```suggestion
```
these seem to be unused

TheBurgerTV2 replied:
sure are! these are leftover from other attempts at making it work, i can get rid of them.

Cyb3RGER commented on BfBBClient.py:
```diff
@@ -1328,6 +1341,55 @@ async def _main(connect, password, patch_file):
+    previous = ctx.previous_rings
+    current_rings = dolphin_memory_engine.read_word(SHINY_COUNT_ADDR)
+
+    if current_rings == 0 and ctx.previous_rings is not None and ctx.previous_rings > 20:
+        # count as death scenario rather
+        pass
```
not quite sure about this check. maybe this can be better checked with ``check_alive`` or the same ``check_control_owner`` check as in ``give_items``?

TheBurgerTV2 replied:
also don't know how to do this lol, coding isn't my strong suit.

Scrungip replied:
For what it's worth, a lot of this was pulled over from Shadow and I kinda haphazardly fit it to work BfBB

Cyb3RGER commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# BfBB AP world
+# BfBB AP world ring link 
```
The ``README.md`` changes need to be reverted so I can merge this.

Cyb3RGER reviewed (changes requested):
Looks good overall but I would like some minor revisions. If you're not sure about something just let me know and will do the changes then.

Cyb3RGER commented on BfBBClient.py:
```diff
@@ -614,6 +621,11 @@ def on_package(self, cmd: str, args: dict):
+            if "tags" in args:
+                related_tags = args["tags"]
+                if "RingLink" in related_tags:
+                    handle_received_rings(self, args["data"])
```
oh one more thing: maybe rename ``handle_received_rings`` to ``on_ringlink`` to match ``on_deathlink``

TheBurgerTV2 replied:
done 🫡

TheBurgerTV2 committed:
Update README.md

TheBurgerTV2 committed:
minor fixes to ring link code

Cyb3RGER committed:
ring/death link changes

- refactoring
- implement /deathlink and /ringlink commands
- implement so-to-ring-ratio

Cyb3RGER committed:
small ring/death link fixes

- properly reset death/ring link even it not present in slot data
- properly reset tags (and password) on disconnect

TheBurgerTV2 committed:
small deathlink message update

Cyb3RGER committed:
death link improvements

- different message based on death/grabbed by hans

Cyb3RGER committed:
ring link changes

- added shiny object to ring ratio yaml options
- added ringlink_ratio client command
